File: apps/ios/WindmillKit/Sources/WindmillGym/RestTimer.swift
import Foundation

public enum Rest {
    // nil is a choice on the dial: no timer.
    public static let choices: [Int?] = [nil, 90, 120, 180]

    public static func target(planEntry: PlanEntry?, preferences: GymPreferences) -> Int? {
        planEntry?.restSeconds ?? preferences.restSeconds
    }

    // The target says where it came from, once, on the timer: a routine entry's own rest names the
    // routine, the dial names nothing. The suffix is the cross-surface unit, byte for byte.
    public static let fromTheRoutine = " · from the routine"

    public static func targetLine(planEntry: PlanEntry?, preferences: GymPreferences) -> String? {
        guard let seconds = target(planEntry: planEntry, preferences: preferences) else { return nil }
        let source = planEntry?.restSeconds == nil ? "" : fromTheRoutine
        return "target " + Readout.clock(Int64(seconds) * 1000) + source
    }

    // A chime later than this is dropped as a false confirmation.
    public static let lateChimeSeconds: Int64 = 5

    // Keyed on both fields: a target changed mid-rest replaces the scheduled sleep.
    public struct Clock: Equatable {
        public let startedAtMs: Int64
        public let targetSeconds: Int

        public init(startedAtMs: Int64, targetSeconds: Int) {
            self.startedAtMs = startedAtMs
            self.targetSeconds = targetSeconds
        }
    }

    public static func filled(targetSeconds: Int, startedAtMs: Int64, now: Int64) -> Double {
        guard targetSeconds > 0 else { return 1 }
        let elapsed = Double(now - startedAtMs) / 1000
        return min(1, max(0, elapsed / Double(targetSeconds)))
    }

    // Time since the last set, counting up throughout: the bar and the chime carry the target, the numeral never flips.
    public static func reading(startedAtMs: Int64, now: Int64) -> String {
        Readout.clock(max(0, now - startedAtMs))
    }
}

File: apps/android/gym/src/main/kotlin/works/windmill/gym/domain/RestTimer.kt
package works.windmill.gym.domain

// Rest is COMPUTED from the instant the set landed, never counted down. The reading counts UP
// throughout — time since the last set, the one reading every surface prints — and a target only adds
// facts beside it: the bar, its name in the label, and the chime. A routine line's `restSeconds`
// outranks the global dial, and the label says so where it does — the timer is the one place on the
// phone that fact is drawn; with no target there is no bar and no sound, and `fraction` is null.

object Rest {
    // The bytes every surface appends when the routine's own line is what the clock runs against.
    const val fromRoutine = " · from the routine"

    data class Target(val seconds: Int, val fromRoutine: Boolean)

    fun target(planEntry: PlanEntry?, preferences: GymPreferences): Target? {
        planEntry?.restSeconds?.let { return Target(it, fromRoutine = true) }
        return preferences.restSeconds?.let { Target(it, fromRoutine = false) }
    }

    class Line(target: Target?, startedAtMs: Long, now: Long) {
        val label: String
        val time: String
        val overrun: Boolean
        val fraction: Float?

        init {
            val elapsed = (now - startedAtMs) / 1000
            time = Readout.clock(elapsed * 1000)
            if (target == null || target.seconds <= 0) {
                overrun = false
                label = "resting"
                fraction = null
            } else {
                overrun = elapsed >= target.seconds
                label = (if (overrun) "rest done · target " else "resting · target ") +
                    Readout.clock(target.seconds * 1000L) +
                    (if (target.fromRoutine) fromRoutine else "")
                // Clamped at both ends: a clock corrected backwards mid-rest draws an empty bar.
                fraction = (elapsed.toFloat() / target.seconds).coerceIn(0f, 1f)
            }
        }
    }
}
